csrc/quantization/vectorization_utils.cuh

```diff
@@ -0,0 +1,51 @@
+// vector_utils.cuh
+#pragma once
+#include "vectorization.cuh"
+
+namespace vllm {
+
+template <typename InT, typename OutT, typename VecOp, typename ScaOp>
+__device__ inline void vectorize_with_alignment(
+    const InT* in, OutT* out, int len, int tid, int stride,
+    VecOp&& vec_op,    // vec_n_t<InT,16> -> vec_n_t<OutT,16>
+    ScaOp&& sca_op) {  // InT -> OutT
+  constexpr int VEC = 16;
+  constexpr int WIDTH = VEC * sizeof(InT);  // 64 B
+  uintptr_t addr = reinterpret_cast<uintptr_t>(in);
+
+  int misalignment_offset = addr & (WIDTH - 1);       // addr % 64
+  int alignment_bytes = WIDTH - misalignment_offset;  // 64 - (addr % 64)
+  int prefix_elems = alignment_bytes & (WIDTH - 1);   // handle 64
+  prefix_elems /= sizeof(InT);
+  prefix_elems = min(prefix_elems, len);  // 0 ≤ prefix < 16
+
+  // 1. prefill the when it is unsafe to vectorize
+  for (int i = tid; i < prefix_elems; i += stride) {
+    sca_op(out[i], in[i]);
```

Comment:
Could you make these return `vout_t`/`OutT` instead of passing in a parameter?

Reply:
Not sure about this, is this functional kernel preferred (return something) or out as a param kernel preferred in vllm?

Reply:
I prefer the return, it's more functional, and I think it's a better general c++/CUDA practice, and the compiler is nominally better at optimizing it (in general - here I think everything will be inlined anyway). So I always default to returning.

Reply:
https://github.com/vllm-project/vllm/issues/19430

I agree with you, and raise another issue for community discussion for a clearer code standard. If the return format is preferred, we can adjust all of the out-pram code in a new pr.